src/Database/LSMTree/Internal.hs

```diff
@@ -558,16 +558,198 @@ updates ::
+
+
+{-# SPECIALISE modifyTableContent_ :: HasFS IO h -> StrictMVar IO (TableContent h) -> (TempRegistry IO (Run (Handle h)) -> TableContent h -> IO (TableContent h)) -> IO () #-}
+-- | Like 'modifyTableContent', but without a return value.
+modifyTableContent_ ::
+     m ~ IO
+  => HasFS m h
+  -> StrictMVar m (TableContent h)
+  -> (TempRegistry m (Run (Handle h)) -> TableContent h -> m (TableContent h))
+  -> m ()
+modifyTableContent_ hfs varContent action =
+    modifyTableContent hfs varContent (\reg content -> (,()) <$> action reg content)
+
+{-# SPECIALISE updatesWithInterleavedFlushes :: TableConfig -> HasFS IO h -> SessionRoot -> UniqCounter IO -> V.Vector (SerialisedKey, Entry SerialisedValue SerialisedBlob) -> TempRegistry IO (Run (Handle h)) -> TableContent h -> IO (TableContent h) #-}
+-- | A single batch of updates can fill up the write buffer multiple times. We
+-- flush the write buffer each time it fills up before trying to fill it up
+-- again.
+updatesWithInterleavedFlushes ::
+     m ~ IO
+  => TableConfig
+  -> HasFS m h
+  -> SessionRoot
+  -> UniqCounter m
+  -> V.Vector (SerialisedKey, Entry SerialisedValue SerialisedBlob)
+  -> TempRegistry m (Run (Handle h))
+  -> TableContent h
+  -> m (TableContent h)
+updatesWithInterleavedFlushes conf hfs root uniqC es0 reg tc = do
+    case updatesWriteBuffer conf es0 (tableWriteBuffer tc) of
+      NotYetFull wb' -> pure $ setWriteBuffer wb' tc
+      Full wb' es' -> do
+        let tc' = setWriteBuffer wb' tc
+        tc'' <- flushWriteBuffer hfs root uniqC reg tc'
+        updatesWithInterleavedFlushes conf hfs root uniqC es' reg tc''
+  where
+    setWriteBuffer :: WriteBuffer -> TableContent h -> TableContent h
+    setWriteBuffer wb' tc0 = TableContent {
+          tableWriteBuffer = wb'
+        , tableLevels = tableLevels tc0
+        , tableCache = tableCache tc0
+        }
+
+-- | The result of a batch of entries to a write buffer.
+data UpdateResult =
+    NotYetFull !WriteBuffer
+  | Full !WriteBuffer !(V.Vector (SerialisedKey, Entry SerialisedValue SerialisedBlob))
+
+-- | Add entries to a write buffer until we run out of entries ('NotYetFull'),
+-- or until the write buffer becomes full ('Full'),
+--
+-- TODO: in case the write buffer is empty, or if it fills up multiple times for
+-- a single batch of updates, we might be able to skip adding entries to the
+-- write buffer for a large part. We can sort and deduplicate the vector of
+-- entries directly, slice it up into properly sized sub-vectors, and write
+-- those to disk.
+--
+-- TODO: should we just flush the write buffer it turns out the batch is so
+-- large that it /might/ fill up the write buffer? If we do so, then we can take
+-- the approach from the previous TODO to flush vectors to disk directly.
+updatesWriteBuffer ::
+     TableConfig
+  -> V.Vector (SerialisedKey, Entry SerialisedValue SerialisedBlob)
+  -> WriteBuffer
+  -> UpdateResult
+updatesWriteBuffer conf es0 wb0 = go es0 wb0
```

Comment:
Probably nicer not to have `es0` and `wb0` in scope for the rest of the function, so could be eta-reduced.

Reply:
Good point